File: leetcode/easy/MinimumDepthOfBinaryTree.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        # If the tree is empty, the depth is 0.
        if not root:
            return 0

        # Use a queue for Breadth-First Search (BFS).
        # Store tuples of (node, current_depth).
        queue = deque([(root, 1)])

        while queue:
            node, depth = queue.popleft()

            # If we find a leaf node, this is the first one we've encountered
            # at the shallowest level, so we can return its depth immediately.
            if not node.left and not node.right:
                return depth

            # Add children to the queue to explore the next level.
            if node.left:
                queue.append((node.left, depth + 1))
            if node.right:
                queue.append((node.right, depth + 1))

File: leetcode/easy/MinimumDepthOfBinaryTree.java
import java.util.ArrayDeque;
import java.util.Deque;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Deque<TreeNode> nodeDeque = new ArrayDeque<>();
        Deque<Integer> depthDeque = new ArrayDeque<>();

        nodeDeque.add(root);
        depthDeque.add(1);

        while (!nodeDeque.isEmpty()) {
            TreeNode node = nodeDeque.poll();
            int depth = depthDeque.poll();

            if (node.left == null && node.right == null) {
                return depth;
            }

            if (node.left != null) {
                nodeDeque.add(node.left);
                depthDeque.add(depth + 1);
            }
            if (node.right != null) {
                nodeDeque.add(node.right);
                depthDeque.add(depth + 1);
            }
        }
        return 0; // Should not be reached
    }
}
